### ISSUE TITLE
Member classes with ES2020 private fields do not respect assignment compatibility

### ISSUE THREAD

**Issue by jmartinezmaes** [microsoft/TypeScript#46375]
# Bug Report

### 🔎 Search Terms

member, class, ES2020, private, field, assignment, compatibility

### 🕗 Version & Regression Information

- This is the behavior in every version I tried starting from 3.8.3.
- I was unable to test this on prior versions because the ES2020 private field syntax was invalid.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=3.8.3#code/C4TwDgpgBAwsD2AnAShYBXRA7AKuCAPDlBAB7ARYAmAzlFhAO5QAUAdBwIaIDmNAXFE5YQAbQC6ASigBeAHxCRCmVGJkK1Og2bsuvAYrFTZCgJZYAZhERQAglAD8dqIIYA3awG4AUN4DGADacNHT2AN7eUFFQiBCcVPBYASBQAEKyUIHBdOlhUADEFvDwAISu6AC2AEbWUAC+vtFQFixVgnBIqBjYeJAEwAAWpjSiAOSpo+Jy0nl+iTTwARBsAfA8LKM88MBQFSmpJaMANFBVkp713g3+8zucAAwZ2nY+c1g0dwCMT0wvvgD0-ygAFFSJA-BQqIoSIhEEg2N5AUJ7mx0sIoZxPqihLEoMBGPAoFRTBYrLEsDssiEIHQLHCKkSSWTKDtzANrKZgMI-NAhtZuH4+TQEQ82C1npjUZIgA)

### 💻 Code

```ts
type CtorReturnType<T extends new (...args: any[]) => any> = T extends new (...args: any[]) => infer A ? A : never;

class A {
    readonly B = class B { #foo!: number }

    f(b: CtorReturnType<this['B']>) { console.log('got my B!', b); }
}

const a0 = new A;
const a1 = new A;

// Expected an error.
// a0.B and a1.B are two different classes from different inheritance hierarchies.
a0.f(new a1.B)
```

### 🙁 Actual behavior

Did not produce an error.

### 🙂 Expected behavior

Expected an error stating that `a1.B` is not assignable to `a0.B`.

Per the handbook:

> Private and protected members in a class affect their compatibility. When an instance of a class is checked for compatibility, if the target type contains a private member, then the source type must also contain a private member that originated from the same class. Likewise, the same applies for an instance with a protected member. This allows a class to be assignment compatible with its super class, **but not with classes from a different inheritance hierarchy which otherwise have the same shape**.

**Comment by sandersn** [comment]
Couple of notes:

1. This should probably apply to `private` in the same way it does to `#`.
2. I modified the above example to get the following JS program:

```js
class A {
    B = class B {
        #foo = 12
        bb() { return ++this.#foo }
    }
    f(b) { console.log('got my', b, this.B.bb.prototype.call(b)) }
}
a0.f(a1.B)
```

That's pretty convincing to me.

**Comment by gibbok** [comment]
@sandersn 
It looks like a bug to me also, I can reproduce it also in this snippet:

```ts

class X {
    public a: string;
    constructor(value: string) {
        this.a = value;
    }
}

class Y extends X {
    public a: string;
    constructor(value: string) {
        super(value)
        this.a = value;
    }
}

class Z {
    public a: string;
    constructor(value: string) {
        this.a = value;
    }
}

let x: X = new X('x')
let y: Y = new Y('y')
let z: Z = new Z('z')
x.a === y.a // valid
z.a === x.a  // valid, but accordingly to the doc this case should be invalid

```

**Comment by RyanCavanaugh** [comment]
This is the same per-invocation class-identity limitation as [#40781](https://github.com/microsoft/TypeScript/issues/40781). TypeScript tracks the single source declaration for the class expression, whereas runtime private-field brands are created for each evaluation. The corresponding `#private` runtime-brand case is confirmed as a design limitation in [#56146](https://github.com/microsoft/TypeScript/issues/56146).
